src/compile/mark/mixins.ts

```diff
@@ -232,31 +229,38 @@ export function wrapCondition(
 }
 
 export function tooltip(model: UnitModel) {
-  const channel = 'tooltip';
-  const channelDef = model.encoding[channel];
+  const {encoding, markDef, config} = model;
+  const channelDef = encoding.tooltip;
   if (isArray(channelDef)) {
-    const keyValues = channelDef.map(fieldDef => {
-      const key = fieldDef.title !== undefined ? fieldDef.title : vgField(fieldDef, {binSuffix: 'range'});
-      const value = ref.text(fieldDef, model.config).signal;
-      return `"${key}": ${value}`;
-    });
-    return {tooltip: {signal: `{${keyValues.join(', ')}}`}};
+    return {tooltip: ref.tooltipForChannelDefs(channelDef, config)};
   } else {
-    // if not an array, behave just like text
-    return textCommon(model, channel, channelDef);
+    return wrapCondition(model, channelDef, 'tooltip', cDef => {
+      // use valueRef based on channelDef first
+      const tooltipRefFromChannelDef = ref.text(cDef, model.config);
+      if (tooltipRefFromChannelDef) {
+        return tooltipRefFromChannelDef;
+      }
+
+      // If tooltipDef does not exist, then use value from markDef or config
+      const markTooltip = markDef.tooltip !== undefined ? markDef.tooltip : getMarkConfig('tooltip', markDef, config);
+      if (isString(markTooltip)) {
+        return {value: markTooltip};
+      } else if (isObject(markTooltip)) {
+        // `tooltip` is `{fields: 'encodings' | 'fields'}`
+        if (markTooltip.fields === 'encodings') {
+          return ref.tooltipForChannelDefs(fieldDefs(encoding), config);
+        } else {
+          return {signal: 'datum'};
```

Comment:
Just to note, we won't use assign here to format the data fields nicer.

Reply:
Yeah, we can consider that later.  Not a priority for now.